### PR TITLE
[MM-48525] AWS SDK v2 migration - IAM

### PR DESCRIPTION
**Pull request by fmartingr**
#### Summary

Upgrade IAM SDK to v2

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-48525

#### Release Note

```release-note
Upgraded AWS SDK for IAM to v2
```

### CODE REVIEW

File: internal/tools/aws/iam.go [hunk 1]
```diff
@@ -30,25 +31,27 @@ type policyStatementEntry struct {
...
 		logger.WithField("iam-user-name", *getResult.User.UserName).Debug("AWS IAM user already created")
 		return getResult.User, nil
 	}
-	if aerr, ok := err.(awserr.Error); ok {
-		if aerr.Code() != iam.ErrCodeNoSuchEntityException {
-			return nil, err
-		}
-	} else {
+	var awsErr *types.NoSuchEntityException
+	if errors.As(err, &awsErr) {
```
Comment (mirshahriar) [inline internal/tools/aws/iam.go:46]: This logic should be opposite. Right?

File: internal/tools/aws/iam.go [hunk 2]
```diff
@@ -140,16 +156,20 @@ func (a *Client) iamEnsureUserDeleted(awsID string, logger log.FieldLogger) erro
...
+		})
 	if err == nil {
 		logger.WithField("iam-policy-name", *getResult.Policy.PolicyName).Debug("AWS IAM policy already created")
 		return getResult.Policy, nil
 	}
-	if aerr, ok := err.(awserr.Error); ok {
-		if aerr.Code() != iam.ErrCodeNoSuchEntityException {
+	if err != nil {
+		var awsErr *types.NoSuchEntityException
+		if errors.As(err, &awsErr) {
```
Comment (mirshahriar) [inline internal/tools/aws/iam.go:172]: This too. Opposite logic.

File: internal/tools/aws/client.go [hunk 3]
```diff
@@ -279,8 +278,8 @@ func (c *Client) buildCloudEnvironmentNameCache() error {
 	}
 
 	for _, alias := range accountAliases.AccountAliases {
-		if strings.HasPrefix(*alias, "mattermost-cloud") && len(strings.Split(*alias, "-")) == 3 {
-			envName := strings.Split(*alias, "-")[2]
+		if strings.HasPrefix(alias, "mattermost-cloud") && len(strings.Split(alias, "-")) == 3 {
```
Comment (mirshahriar) [inline internal/tools/aws/client.go:281]: Its splitting twice. You can address this too. 0/5

**Review by mirshahriar — CHANGES_REQUESTED** [review]

**Review by mirshahriar — APPROVED** [review]
LGTM

**Review by gabrieljackson — APPROVED** [review]

File: internal/tools/aws/iam.go [hunk 4]
```diff
@@ -140,16 +156,20 @@ func (a *Client) iamEnsureUserDeleted(awsID string, logger log.FieldLogger) erro
...
+		&iam.GetPolicyInput{
+			PolicyArn: aws.String(policyARN),
+		})
 	if err == nil {
 		logger.WithField("iam-policy-name", *getResult.Policy.PolicyName).Debug("AWS IAM policy already created")
 		return getResult.Policy, nil
 	}
-	if aerr, ok := err.(awserr.Error); ok {
-		if aerr.Code() != iam.ErrCodeNoSuchEntityException {
+	if err != nil {
```
Comment (mirshahriar) [inline internal/tools/aws/iam.go:170]: We do not need this `if/else` check anymore

**Review by mirshahriar — CHANGES_REQUESTED** [review]

File: internal/tools/aws/iam.go [hunk 5]
```diff
@@ -140,16 +158,20 @@ func (a *Client) iamEnsureUserDeleted(awsID string, logger log.FieldLogger) erro
...
+		var awsErr *types.NoSuchEntityException
+		if errors.As(err, &awsErr) {
```
Comment (mirshahriar) [inline internal/tools/aws/iam.go]: This will be opposite.

**Review by mirshahriar — CHANGES_REQUESTED** [review]

**Review by mirshahriar — APPROVED** [review]

File: internal/tools/aws/iam.go (hunk 4 above)
Comment (fmartingr) [reply internal/tools/aws/iam.go]: Please review this simplified logic change: https://github.com/mattermost/mattermost-cloud/pull/803/commits/8e02ca327ea0a08b0d6a458b4b2e3c62a22014f2

**Review by mirshahriar — APPROVED** [review]

**Pull request merged**